plugins/mtd/fu-uboot-firmware.c

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (C) 2022 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <fwupdplugin.h>
+
+#include "fu-uboot-firmware.h"
+
+struct _FuUbootFirmware {
+	FuFirmware parent_instance;
+};
+
+G_DEFINE_TYPE(FuUbootFirmware, fu_uboot_firmware, FU_TYPE_FIRMWARE)
+
+static gboolean
+fu_uboot_firmware_parse(FuFirmware *firmware,
+			GBytes *fw,
+			guint64 addr_start,
+			guint64 addr_end,
+			FwupdInstallFlags flags,
+			GError **error)
+{
+	const gchar needle1[] = {"U-Boot TPL "};
+	const gchar needle2[] = {" "};
+	gsize offset1 = 0;
+	gsize offset2 = 0;
+	g_autofree gchar *version = NULL;
+
+	/* find start and end */
+	if (!fu_memmem_safe(g_bytes_get_data(fw, NULL),
+			    g_bytes_get_size(fw), /* haystack */
+			    (const guint8 *)needle1,
+			    sizeof(needle1) - 1,
+			    &offset1,
+			    error))
+		return FALSE;
+	offset1 += sizeof(needle1) - 1;
+	if (!fu_memmem_safe((const guint8 *)g_bytes_get_data(fw, NULL) + offset1,
+			    g_bytes_get_size(fw) - offset1, /* haystack */
+			    (const guint8 *)needle2,
+			    sizeof(needle2) - 1,
+			    &offset2,
+			    error))
+		return FALSE;
+
+	/* set version */
+	version = g_strndup((const gchar *)g_bytes_get_data(fw, NULL) + offset1, offset2);
+	fu_firmware_set_version(firmware, version);
+
+	/* success */
+	return TRUE;
+}
+
+static GBytes *
+fu_uboot_firmware_write(FuFirmware *firmware, GError **error)
```

Comment:
I cannot follow here, are we flashing something different than the binary here?

Reply:
Ya, this is for hypothetical fuzzing support which I didn't add yet. This allows us to say "write me a UBOOT firmware that looks plausible with this fake data" to allow the fuzzer to do it's thing. If you've got any more info on what the uboot image actually looks like (e.g. offsets and stuff) it makes sense to properly parse the image rather than doing a dumb `memmem()` on a specific string.

Reply:
Actually; on this; rather than doing the memmem thing; what are we looking at here? It's not the FIT as that's later in the image I think. Is this the SPL that contains the U-Boot version string? i.e. what blob am I seeing at 0x15000 from the SPI dump? Is there a header that specifies the jump to 0x15000 to avoid us reading in a ton of stuff we won't need?

Reply:
I've moved this out to another PR, we can discuss there. I'd still like to know what the layout is. I'm also erring on looking for the *pinephone* firmware version (e.g. a custom firmware header, embedded in the firmware somewhere) rather than a *uboot* version.

Reply:
@hughsie I will be able to dig up more soon, currently swamped with $DAYJOB deadlines. I opened the issue to discuss a bit, didn't expect that things would be moving so fast :joy: I will invite the Tow-Boot maintainer to have a look, he knows more.

Reply:
> didn't expect that things would be moving so fast

Ha! It's **me**, what did you expect :)

Reply:
Yeah true, you're so fast in fixing things, awesome!